### DISCUSSION TITLE
How does Cadence perform client version compatibility check

### DISCUSSION THREAD

**Discussion by zhangkyou** [uber/cadence#4369] [Q&A] [answered]
我们的cadence server部署的是0.20.0版本，然后cadence client中import的cadence是0.16.0，这两边的cadence版本不一致会有什么问题吗？

**Comment by longquanzheng** [comment]
By convention, Cadence server is always compatible to old clients, but may not for newer clients. 

Cadence is using "SDK feature version" to compare, and perform the compatibility check if you enable the "ClientVersionCheck" feature by dynamic config:

```
frontend.enableClientVersionCheck:
  - value: true
```
This feature is optional. You can enable it if you want to get some early protection. 
Nothing will go seriously wrong if you don't enable this feature. It's mainly for some feature that supported by background task so that the error is not obvious. For example, delay start workflow, batcher system workflow, etc. 

To look at feature versions, for examples:

This is the highest feature version that 0.21.3 server support for go client：
https://github.com/uber/cadence/blob/v0.21.3/common/client/versionChecker.go#L43

This is feature version of 0.16.0 go client :
https://github.com/uber-go/cadence-client/blob/v0.16.0/internal/version.go#L45

3.2.0 Java client feature version is defined
https://github.com/uber/cadence-java-client/blob/v3.2.0/src/main/java/com/uber/cadence/internal/Version.java#L46

For other versions of servers/clients, you can replace the tag in the URL to see it.
In the future, we will show the feature version information in the binary as well after this PR is released: #4308

========Chinese version===============
server 0.20.0 对应go client 0.16.0是完全没有问题的。
这个是0.20.0 server支持的go client的feature version：
https://github.com/uber/cadence/blob/v0.20.0/common/client/versionChecker.go#L43

这个是0.16.0 go client 的feature version:
https://github.com/uber-go/cadence-client/blob/v0.16.0/internal/version.go#L45

都是1.6，所以是完全兼容.

首先，sever会向下兼容老的client，但可能会没法支持新的client.
但即便是新的client ，默认也是会尽量兼容。然而有些情况无法兼容，比如说新的client 需要server 实现新的feature . 这种情况下使用新的feature 就会在使用过程中报错。

如果你需要让这个错误更明显一些，需要在server 开启feature version checking. 默认是关闭的。开启后如果client 太新，会报类似的错误https://stackoverflow.com/questions/68217385/what-is-clientversionnotsupportederror-and-how-to-resolve-it

**Comment by longquanzheng** [comment, marked as answer]
Similarly 

3.2.0 Java client feature version is defined 
https://github.com/uber/cadence-java-client/blob/v3.2.0/src/main/java/com/uber/cadence/internal/Version.java#L46

0.20.0 server supported java client feature version is defined in https://github.com/uber/cadence/blob/v0.20.0/common/client/versionChecker.go#L45

For other versions of servers/clients, you can replace the tag in the URL to see it. 
In the future, we will show the feature version information in the binary as well after this PR is released: https://github.com/uber/cadence/pull/4308
